guide/src/async-await.md

```diff
@@ -0,0 +1,240 @@
+an `async` function is _just like calling a regular function_, the only difference is that we have
+to do something special with the object that it returns.
+
+Normally in Python, that something special is the `await` keyword, but in Rust, we don't have the 
+luxury of using Python's syntax. Luckily, Rust also has an `await` keyword that does something
+similar, we just need to find a way of converting a `&PyAny` into a Rust future so we can use the
+`await` keyword on it.
+
+That's where pyo3-asyncio comes in. `pyo3_asyncio::into_future` performs this conversion for us:
+
+```rust
+let future = Python::with_gil(|py| {
+    // import the module containing the py_sleep function
+    let example = py.import("example")?;
+
+    // calling the py_sleep method like a normal function returns a coroutine
+    let coroutine = example.call_method0("py_sleep")?;
+
+    // convert the coroutine into a Rust future
+    pyo3_asyncio::into_future(py, coroutine)
+})?;
+
+// await the future
+future.await;
+```
+
+## Awaiting a Rust Future in Python
+
+Here we have the same async function as before written in Rust using the 
+[`async-std`](https://async.rs/) runtime:
+
+```rust
+async fn rust_sleep() {
+    async_std::task::sleep(Duration::from_secs(1)).await;
+}
+```
+
+Similar to Python, Rust's async functions also return a special object called a
+`Future`:
+
+```rust
+let future = rust_sleep();
+```
+
+We can convert this future object into Python to make it "Awaitable". This tells Python that you can
+use the `await` keyword with it. In order to do this, we'll call 
+`pyo3_asyncio::async_std::into_coroutine`:
+
+```rust
+#[pyfunction]
+fn call_rust_sleep(py: Python) -> PyResult<PyObject> {
+    pyo3_asyncio::async_std::into_coroutine(py, async move {
+        rust_sleep().await;
+        Ok(())
+    })
+}
+```
+
+In Python, we can call this pyo3 function just like any other async function:
+
+```python
+from example import call_rust_sleep
+
+async def rust_sleep():
+    await call_rust_sleep()
+```
+
+## Managing Event Loops
+
+Python's event loop requires some special treatment, especially regarding the main thread. Some of
+Python's Asyncio features, like proper signal handling, require control over the main thread, which
+doesn't always play well with Rust.
+
+Luckily, Rust's event loops are pretty flexible and don't _need_ control over the main thread, so in
+pyo3-asyncio, we decided the best way to handle Rust/Python interop was to just surrender the main
+thread to Python and run Rust's event loops in the background. Unfortunately, since most event loop 
+implementations _prefer_ control over the main thread, this can still make some things awkward.
+
+### PyO3 Asyncio Initialization
+
+Because Python needs to control the main thread, we can't use the convenient proc macros from Rust
+runtimes to handle the main function or test functions. 
+
+Instead, the initialization for PyO3 has to be done manually from the main function and the main 
```

Comment:
I haven't been following the conversation on the `auto-initialize`, so I'm not sure what I should add here. AFAIK there are no problems with the current implementation using the `auto-initialize` feature. Are you suggesting that we add an example of what someone should do if they want to disable `auto-initialize`?

Signal handling _appears_ to work properly in my project that uses pyo3-asyncio, so that difference between `Py_InitializeEx(0)` and `Py_InitializeEx(1)` kinda surprises me. Maybe that's something we should look at a bit further.